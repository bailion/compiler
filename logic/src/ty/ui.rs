use codespan_reporting::{
    files::SimpleFiles,
    term::{self, termcolor::Buffer, Config},
};

use crate::{id::tag, parse::parse, ty::type_check};

fn setup_file(file: String) -> (usize, SimpleFiles<String, String>) {
    let mut files = SimpleFiles::new();
    let id = files.add("file".to_string(), file);
    (id, files)
}

fn ui_test(string: String) {
    let tree = parse(&string).unwrap();
    let tagged = tag(tree);
    let err = type_check(&tagged).unwrap_err();
    let (file_id, files) = setup_file(string.clone());
    let report = err.report(file_id);

    let mut writer = Buffer::no_color();

    let config = Config::default();

    term::emit(&mut writer, &config, &files, &report).unwrap();

    let vec = writer.into_inner();
    let string = String::from_utf8(vec).unwrap();

    insta::assert_snapshot!(string);
}

#[test]
fn test_simple() {
    ui_test(include_str!("examples/failing").to_string());
}
