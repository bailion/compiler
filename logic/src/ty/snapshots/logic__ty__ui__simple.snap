---
source: logic/src/ty/ui.rs
expression: "ui_test(include_str!(\"examples/failing\").to_string())"
---
error: Your program contains a type error!
  ┌─ file:3:12
  │
3 │     return "Hello World!"
  │            ^^^^^^^^^^^^^^ this is of type StrSlice
4 │   else
5 │     return 1
  │            ^ this is of type `Int` which is not the same as type `StrSlice`


